File: src/algorithms/arrays/rotate-array/sources/RotateArray.java
// Rotate Array (Reversal Method) — O(n) three-reversal technique with O(1) space
public class RotateArray {
    public static int[] rotateArray(int[] inputArray, int rotateCount) {
        int[] result = inputArray.clone(); // @step:initialize
        int arrayLength = result.length; // @step:initialize

        if (arrayLength == 0) { // @step:initialize
            return result; // @step:initialize
        }

        int effectiveRotation = rotateCount % arrayLength; // @step:initialize

        if (effectiveRotation == 0) { // @step:initialize
            return result; // @step:initialize
        }

        // Phase 1: reverse entire array
        reverseSegment(result, 0, arrayLength - 1); // @step:swap

        // Phase 2: reverse first effectiveRotation elements
        reverseSegment(result, 0, effectiveRotation - 1); // @step:swap

        // Phase 3: reverse remaining elements
        reverseSegment(result, effectiveRotation, arrayLength - 1); // @step:swap

        return result; // @step:complete
    }

    private static void reverseSegment(int[] arr, int leftPointer, int rightPointer) {
        while (leftPointer < rightPointer) {
            int tempValue = arr[leftPointer]; // @step:swap
            arr[leftPointer] = arr[rightPointer]; // @step:swap
            arr[rightPointer] = tempValue; // @step:swap
            leftPointer++; // @step:visit
            rightPointer--; // @step:visit
        }
    }
}

File: src/algorithms/arrays/rotation/rotate-array/sources/rotate-array.py
# Rotate Array (Reversal Method) — O(n) three-reversal technique with O(1) space
def rotate_array(input_array: list[int], rotate_count: int) -> list[int]:
    result = input_array[:]
    array_length = len(result)

    if array_length == 0:  # @step:initialize
        return result  # @step:initialize

    effective_rotation = rotate_count % array_length  # @step:initialize

    if effective_rotation == 0:  # @step:initialize
        return result  # @step:initialize

    def reverse_segment(arr: list[int], left: int, right: int) -> None:
        while left < right:
            arr[left], arr[right] = arr[right], arr[left]  # @step:swap
            left += 1  # @step:visit
            right -= 1  # @step:visit

    # Phase 1: reverse entire array
    reverse_segment(result, 0, array_length - 1)  # @step:swap

    # Phase 2: reverse first effective_rotation elements
    reverse_segment(result, 0, effective_rotation - 1)  # @step:swap

    # Phase 3: reverse remaining elements
    reverse_segment(result, effective_rotation, array_length - 1)  # @step:swap

    return result  # @step:complete
